### ISSUE TITLE
Issue 004: `src` 非変更で Coarse train logits に `velocity_mask` を適用する

### ISSUE THREAD

**Issue by ozw4** [ozw4/seisai#289]
# Issue 004: `src` 非変更で Coarse train logits に `velocity_mask` を適用する

## 背景

ユーザー要件では、速度フィルタは既存の以下を使う。

```text
/workspace/packages/seisai-engine/src/seisai_engine/postprocess/velocity_filter_op.py
```

ただし `ApplyVelocityFiltOp.__call__()` は `@torch.no_grad()` 付きなので、学習 loss の前にそのまま呼ぶと logits への勾配が切れる。Coarse 学習で使う場合は、同じ file 内の lower-level function を使って mask を作り、勾配を保ったまま logits に加算する必要がある。

この issue では、Issue 003 の site54 train runner / criterion hook に `velocity_mask` を実装する。

## 重要な制約

- `packages/seisai-engine/src/**` は変更しない。
- `packages/seisai-pick/src/**` は変更しない。
- 速度 mask の生成・適用は既存 `velocity_filter_op.py` の関数を import して使う。
- `ApplyVelocityFiltOp` 本体は学習 loss 前には使わない。`@torch.no_grad()` により勾配が切れるため。
- baseline / `velocity_mask.enabled: false` では挙動を変えない。
- fold list、train / inner_valid / heldout split は変更しない。

## 現在の repo 前提

既存関数:

```python
from seisai_engine.postprocess.velocity_filter_op import (
    make_velocity_feasible_filt,
    apply_velocity_filt_logits,
)
```

`make_velocity_feasible_filt()` は以下を受け取る。

```text
offsets_m: (B,H) [m]
dt_sec:    (B,) / (B,1) / (B,1,1) / scalar [s]
W:         time bins
vmin, vmax, t0_lo_ms, t0_hi_ms, taper_ms
```

Coarse dataset の batch には以下が入る。

```text
batch["offsets"]       # (B,H), anchor offsets [m]
batch["dt_sec"]        # effective Coarse dt [s]
batch["fb_idx_coarse"] # (B,H), Coarse label index; trend issue で使用
```

## 目的

Coarse training の base KL loss 前に、experiment config で指定した場合だけ velocity feasible mask を logits に適用する。

概念的には以下。

```python
pred = model(x)
pred_for_loss = apply_site54_velocity_mask_if_enabled(pred, batch, cfg)
base_loss = base_criterion(pred_for_loss, target, batch)
```

## 変更対象

主に Issue 003 で追加された runtime files を変更する。

```text
proc/fbpick/site54/oof/scripts/site54_coarse_experiment_runtime.py
proc/fbpick/site54/oof/scripts/run_fbpick_coarse_train_site54_experiment.py
```

必要なら以下も更新する。

```text
proc/fbpick/site54/oof/experiments/README.md
proc/fbpick/site54/oof/README.md
```

変更禁止:

```text
packages/seisai-engine/src/**
packages/seisai-pick/src/**
```

## config interface

`site54_oof_experiment.coarse_train_runtime.velocity_mask` を読む。

```yaml
site54_oof_experiment:
  coarse_train_runtime:
    enabled: true
    velocity_mask:
      enabled: true
      apply_train: true
      apply_inner_valid: false
      vmin: 300.0
      vmax: 6000.0
      t0_lo_ms: -20.0
      t0_hi_ms: 200.0
      taper_ms: 8.0
      power: 1.0
      offsets_key: offsets
      dt_key: dt_sec
      channels: 0
      ensure_nonempty: true
      aux_key: velocity_mask
```

### default 値の考え方

- baseline では section が無い、または `enabled: false`。
- 初期 site54 experiment では、repo 既存の `make_physics_fold_configs.py` にある `physical_band` と同じ値を使う。

```yaml
vmin: 300.0
vmax: 6000.0
t0_lo_ms: -20.0
t0_hi_ms: 200.0
```

- `taper_ms` と `power` は `velocity_filter_op.VelocityFiltConfig` の既存 default に合わせる。

```yaml
taper_ms: 8.0
power: 1.0
```

### 1. config parsing

local helper を追加する。dataclass でも dict helper でもよい。

必須 validation:

- `vmin > 0`
- `vmax >= vmin`
- `taper_ms >= 0`
- `power > 0`
- `channels` は int または int list
- `offsets_key` と `dt_key` は non-empty str

### 2. mask 生成

```python
mask = make_velocity_feasible_filt(
    offsets_m=batch_on_device[offsets_key],
    dt_sec=batch_on_device[dt_key],
    W=logits.shape[-1],
    vmin=vmin,
    vmax=vmax,
    t0_lo_ms=t0_lo_ms,
    t0_hi_ms=t0_hi_ms,
    taper_ms=taper_ms,
    device=logits.device,
    dtype=torch.float32,
)
```

batch の `offsets` / `dt_sec` は logits と同じ device に移す。元 batch は破壊しない。

### 3. selected channels のみに適用

`apply_velocity_filt_logits(logits, mask, power=power)` は全 channel に broadcast 適用する。Coarse は output channel 1 なので `channels: 0` が基本だが、実装は selected channel のみに適用できるようにする。

推奨:

```python
out = logits.clone()
masked_all = apply_velocity_filt_logits(logits, mask, power=power)
for ch in normalized_channels:
    out[:, ch, :, :] = masked_all[:, ch, :, :]
return out
```

### 4. 勾配

- `make_velocity_feasible_filt()` の mask は batch geometry 由来なので勾配不要。
- `apply_velocity_filt_logits()` は `logits + log(mask)` なので logits への勾配は保たれる。
- `ApplyVelocityFiltOp.__call__()` は `@torch.no_grad()` なので使用しない。

### 5. `ensure_nonempty`

各 `(B,H)` trace で mask sum が 0 の場合、`apply_velocity_filt_logits()` は全 time bin を `-inf` にし得る。`ensure_nonempty: true` の場合は、その trace の mask を all ones にフォールバックする。

```python
empty = mask.sum(dim=-1) <= 0
mask = torch.where(empty.unsqueeze(-1), torch.ones_like(mask), mask)
```

この fallback の件数は aux / runtime summary に残す。

### 6. train loss への適用

`apply_train: true` のとき、base criterion の前に logits を mask する。

```python
pred_for_loss, velocity_aux = apply_velocity_mask(pred, batch, cfg)
base_loss = base_criterion(pred_for_loss, target, batch)
```

`apply_train: false` なら何もしない。

### 7. inner validation への適用

`apply_inner_valid: true` のとき、inner validation の `infer_loss` と debug PNG にも mask 後 logits を使う。

比較のため、初期 default は `false` 推奨。これにより checkpoint metric は従来の unmasked `infer_loss` と比較しやすい。

### 8. diagnostics

training loop の各 step に詳細ログを出す必要はない。必要なら runtime summary に以下を保存する。

```json
{
  "velocity_mask_enabled": true,
  "velocity_mask_apply_train": true,
  "velocity_mask_apply_inner_valid": false,
  "velocity_mask_empty_trace_count_last": 0
}
```

## 数値上の注意

既存 `apply_velocity_filt_logits()` は mask=0 を `-inf` にする。label の Gaussian support が完全に mask 外にある場合、base KL loss が非有限になる可能性がある。

実装では以下のどちらかを行う。

1. 既存関数の挙動を尊重して、loss が非有限になったら明確な error message を出す。
2. 追加 option `finite_log_floor` を用意し、明示設定された場合だけ `log(mask).clamp_min(finite_log_floor)` 相当にする。

default は既存関数と同じ挙動にし、勝手に finite floor を有効化しない。

## 受け入れ条件

- `packages/seisai-engine/src/**` と `packages/seisai-pick/src/**` に変更がない。
- `velocity_mask.enabled: false` では baseline no-op。
- `velocity_mask.enabled: true, apply_train: true` で base criterion 前の logits が mask される。
- logits への gradient が残ることを小さい unit test で確認する。
- `ApplyVelocityFiltOp.__call__()` を train loss 前に使っていない。
- `offsets` / `dt_sec` が CPU batch でも CUDA logits と同じ device に移される。
- `ensure_nonempty` fallback が動く。

## 確認コマンド

```bash
cd /workspace

python -m py_compile \
  proc/fbpick/site54/oof/scripts/site54_coarse_experiment_runtime.py \
  proc/fbpick/site54/oof/scripts/run_fbpick_coarse_train_site54_experiment.py

python - <<'PY'
import torch
from pathlib import Path
import sys
sys.path.insert(0, str(Path('proc/fbpick/site54/oof/scripts').resolve()))
from site54_coarse_experiment_runtime import apply_velocity_mask_to_logits

logits = torch.randn(2, 1, 4, 32, requires_grad=True)
batch = {
    'offsets': torch.tensor([[0., 100., 200., 300.], [0., 100., 200., 300.]]),
    'dt_sec': torch.tensor([0.004, 0.004]),
}
cfg = {
    'enabled': True,
    'vmin': 300.0,
    'vmax': 6000.0,
    't0_lo_ms': -20.0,
    't0_hi_ms': 200.0,
    'taper_ms': 8.0,
    'power': 1.0,
    'offsets_key': 'offsets',
    'dt_key': 'dt_sec',
    'channels': 0,
    'ensure_nonempty': True,
}
out, aux = apply_velocity_mask_to_logits(logits, batch, cfg)
loss = torch.nan_to_num(out, neginf=-30.0, posinf=30.0).sum()
loss.backward()
assert logits.grad is not None
print('ok', aux.keys())
PY
```
